hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/slowlog/SlowLogEventHandler.java

```diff
@@ -0,0 +1,167 @@
+ * limitations under the License.
+ */
+
+package org.apache.hadoop.hbase.regionserver.slowlog;
+
+import com.lmax.disruptor.EventHandler;
+import com.lmax.disruptor.RingBuffer;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.Queue;
+import java.util.stream.Collectors;
+
+import org.apache.commons.lang3.StringUtils;
+import org.apache.hadoop.hbase.client.SlowLogParams;
+import org.apache.hadoop.hbase.util.Pair;
+import org.apache.yetus.audience.InterfaceAudience;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import org.apache.hbase.thirdparty.com.google.common.collect.EvictingQueue;
+import org.apache.hbase.thirdparty.com.google.common.collect.Queues;
+import org.apache.hbase.thirdparty.com.google.protobuf.Message;
+
+import org.apache.hadoop.hbase.shaded.protobuf.ProtobufUtil;
+import org.apache.hadoop.hbase.shaded.protobuf.generated.AdminProtos;
+import org.apache.hadoop.hbase.shaded.protobuf.generated.TooSlowLog.SlowLogPayload;
+
+/**
+ * Event Handler run by disruptor ringbuffer consumer
+ */
+@InterfaceAudience.Private
+class SlowLogEventHandler implements EventHandler<RingBufferEnvelope> {
+
+  private static final Logger LOG = LoggerFactory.getLogger(SlowLogEventHandler.class);
+
+  private final Queue<SlowLogPayload> queue;
+
+  SlowLogEventHandler(int eventCount) {
+    EvictingQueue<SlowLogPayload> evictingQueue = EvictingQueue.create(eventCount);
+    queue = Queues.synchronizedQueue(evictingQueue);
+  }
+
+  /**
+   * Called when a publisher has published an event to the {@link RingBuffer}
+   *
+   * @param event published to the {@link RingBuffer}
+   * @param sequence of the event being processed
+   * @param endOfBatch flag to indicate if this is the last event in a batch from
+   *   the {@link RingBuffer}
+   * @throws Exception if the EventHandler would like the exception handled further up the chain
+   */
+  @Override
+  public void onEvent(RingBufferEnvelope event, long sequence, boolean endOfBatch)
+      throws Exception {
+    Pair<SlowLogPayload, Message> slowLogPayloadPair = event.getPayload();
+    SlowLogPayload slowLogPayload = slowLogPayloadPair.getFirst();
+    Message param = slowLogPayloadPair.getSecond();
+    if (!slowLogPayload.hasParam() || !slowLogPayload.hasRegionName()) {
+      SlowLogParams slowLogParams = ProtobufUtil.getSlowLogParams(param);
+      slowLogPayload = slowLogPayload.toBuilder()
+        .setParam(slowLogParams != null ? slowLogParams.getParams() : StringUtils.EMPTY)
+        .setRegionName(slowLogParams != null ? slowLogParams.getRegionName() : StringUtils.EMPTY)
+        .build();
+    }
+    if (LOG.isTraceEnabled()) {
+      LOG.trace("Received Slow Log Event. RingBuffer sequence: {}, isEndOfBatch: {}, " +
+          "Event Call: {}", sequence, endOfBatch,
+        slowLogPayload.getCallDetails());
+    }
+    queue.add(slowLogPayload);
```

Comment:
Like I mentioned elsewhere, this should not be a Queue of protobuf serialized stuff, instead it should be a simple container class that is serialized as output only when there is a request to dump it.